### ISSUE TITLE
Fix Operations API on Kube: Refactor normalization transform_config into Java

### ISSUE THREAD

**Issue by davinchia** [airbytehq/airbyte#5091]
- **Severity**: High

## Current Behavior
- See this [slack thread](https://airbytehq.slack.com/archives/C021JANJ6TY/p1627506369378200) for info.
- The CustomDbtRunner no longer works on Kube since it assumes processes are able to share file space. This is not true for Kube processes.
- What happens is the first process is started to prepare certain files DBT requires. The second process is then started up to run DBT, however the files are missing and it errors out.

There are two approaches here: 
1) Make the user installed git and the base normalize folder in their submitted docker image
2)  Migrate the transform_config directory to Java. This way the scheduler can run this and transfer the yml file over to the container. 
    * Submitted image will still need git
    * need to modify normalization as well (all we need to do is remove this from the entrypoint.sh in base-normalization and make sure we also copy the yaml over)

We are choosing option 2 since it's simpler to implement.

## Expected Behavior
This should work on Kube.

**Comment by jrhizor** [comment]
relevant slack threads: 
- https://airbytehq.slack.com/archives/C021JANJ6TY/p1627506369378200
- https://airbytehq.slack.com/archives/C01MFR03D5W/p1629122790159600?thread_ts=1628871489.083000&cid=C01MFR03D5W

**Comment by Phlair** [comment]
this will also need to take into account [ssh tunnelling process](https://github.com/airbytehq/airbyte/pull/5818)

**Comment by zackliston** [comment]
This is blocking all usage for custom DBT transformations in kubernetes deployments. What's the expected timeline? Are there workarounds that could be shared?

**Comment by ro8inmorgan** [comment]
Wish I read this before going through the hassle of setting up a kube install. Now I waisted my morning to this. This is a serious feature breaking bug

**Comment by RajuGujjalapatiChowdary** [comment]
My dbt is not working on airbyte(deployed in kube).
Any solutions on this?

**Comment by haithem-souala** [comment]
Any ETA?

**Comment by nathanaevitas** [comment]
This issue gonna celebrate its first birthday to be 1 year old.

**Comment by GabrielGLevine** [comment]
Is anyone aware of a workaround for this?

**Comment by evantahler** [comment]
Alternative suggestion - Approach #3: shared temporary directory.
We could use something like Fuse+S3 mounts to make the directory in which the first process prepares the dbt files backed by a "network drive", then available to the second process. We would need a way to pass configuration in for the S3 mount (e.g. a destination config option), or set via ENV

**Comment by stegacha** [comment]
Any ETA? or some workaround?

**Comment by alepietrobon** [comment]
+1

**Comment by ThoSap** [comment]
+1

**Comment by davisford** [comment]
Hi airbyte team.  This seems to be a serious limitation right now.  If someone could point me to any particular documentation on how airbyte bootstraps dbt and how it works, src is fine too, I would be happy to attempt to dig into a solution.  We are interested in using both technologies together in k8s, and this is obviously a show-stopper issue.  It's a bit concerning it has been out there for a year now :(

**Comment by davinchia** [comment]
Hi all, sorry for the slow reply here! We know this is a hole and are trying to fit this into our list.

@davisford this is complicated enough I'm not sure we want an OSS member to contribute the fix. Will discuss and get back to you.

fyi @malikdiarra @pmossman @andyjih

**Comment by davisford** [comment]
Thanks @davinchia -- I realize there are multiple ways to approach resolving this issue.  I wasn't necessarily promoting the idea of adjusting the airbyte java src -- (although I probably could), but rather thinking of other possible workarounds to get people past this.  I know k8s quite well.  I'm curious if we could define a persistent volume / persistent volume claim that is attached to all pods in question it wouldn't resolve the issue.  I just don't quite understand the architecture of how airbyte tries to orchestrate / spawn the dbt transforms well enough yet to figure out if this could resolve things.  I'm still reading through the source to see.

**Comment by GabrielGLevine** [comment]
@davinchia It seems like the initial root of the issue was the inability for Kube processes to share file space:
> The CustomDbtRunner no longer works on Kube since it assumes processes are able to share file space. This is not true for Kube processes.
I can't speak for all platforms but I know that GKE at least now supports ReadWriteMany behavior via their filestore driver. I believe AWS EFS could accomplish something similar but I'm not as familiar.

**Comment by betocolon23** [comment]
Any updates on this? 😭

**Comment by GabrielGLevine** [comment]
@davinchia I saw this was on the roadmap for Q4 2022 but it seems to have fallen off the official roadmap: https://app.harvestr.io/roadmap/view/pQU6gdCyc/launch-week-roadmap. Is work on this by the core team still planned?

**Comment by GabrielGLevine** [comment]
I'm working on implementing the [workaround](https://airbytehq.slack.com/archives/C021JANJ6TY/p1627556825392100?thread_ts=1627506369.378200&cid=C021JANJ6TY) @davinchia suggested to build a Docker container with the dbt code built into the image to avoid needing to read it from disk, but I'm running into some execution errors. Would appreciate any advice on how to configure this and hopefully it can serve as a guide to others working around.

Here's my Dockerfile config:
```
FROM ghcr.io/dbt-labs/dbt-postgres:1.2.3

COPY dbt_project .

ENV DB_USER=
ENV DB_PASS=
ENV DB_HOST=

RUN dbt deps --project-dir data_proxy --profiles-dir .dbt --target proxy_db

CMD ["dbt", "build", "--project-dir", "data_proxy", "--profiles-dir", ".dbt", "--target", "proxy_db"]
```

Logs:
```
	at io.airbyte.workers.temporal.TemporalAttemptExecution.get(TemporalAttemptExecution.java:161) ~[io.airbyte-airbyte-workers-0.40.25.jar:?]
	at io.airbyte.workers.temporal.sync.DbtTransformationActivityImpl.lambda$run$3(DbtTransformationActivityImpl.java:132) ~[io.airbyte-airbyte-workers-0.40.25.jar:?]
	at io.airbyte.commons.temporal.TemporalUtils.withBackgroundHeartbeat(TemporalUtils.java:276) ~[io.airbyte-airbyte-commons-temporal-0.40.25.jar:?]
	... 14 more
Caused by: io.airbyte.workers.exception.WorkerException: Dbt Transformation Failed.
	at io.airbyte.workers.general.DbtTransformationWorker.run(DbtTransformationWorker.java:72) ~[io.airbyte-airbyte-commons-worker-0.40.25.jar:?]
	at io.airbyte.workers.general.DbtTransformationWorker.run(DbtTransformationWorker.java:26) ~[io.airbyte-airbyte-commons-worker-0.40.25.jar:?]
	at io.airbyte.workers.temporal.TemporalAttemptExecution.lambda$getWorkerThread$5(TemporalAttemptExecution.java:195) ~[io.airbyte-airbyte-workers-0.40.25.jar:?]
	... 1 more
Caused by: io.airbyte.workers.exception.WorkerException: DBT Transformation Failed.
	at io.airbyte.workers.general.DbtTransformationWorker.run(DbtTransformationWorker.java:68) ~[io.airbyte-airbyte-commons-worker-0.40.25.jar:?]
	at io.airbyte.workers.general.DbtTransformationWorker.run(DbtTransformationWorker.java:26) ~[io.airbyte-airbyte-commons-worker-0.40.25.jar:?]
	at io.airbyte.workers.temporal.TemporalAttemptExecution.lambda$getWorkerThread$5(TemporalAttemptExecution.java:195) ~[io.airbyte-airbyte-workers-0.40.25.jar:?]
	... 1 more
	Suppressed: io.airbyte.workers.exception.WorkerException: Normalization process did not terminate normally (exit code: 128)
		at io.airbyte.workers.normalization.DefaultNormalizationRunner.close(DefaultNormalizationRunner.java:217) ~[io.airbyte-airbyte-commons-worker-0.40.25.jar:?]
		at io.airbyte.workers.general.DbtTransformationRunner.close(DbtTransformationRunner.java:128) ~[io.airbyte-airbyte-commons-worker-0.40.25.jar:?]
		at io.airbyte.workers.general.DbtTransformationWorker.run(DbtTransformationWorker.java:57) ~[io.airbyte-airbyte-commons-worker-0.40.25.jar:?]
		at io.airbyte.workers.general.DbtTransformationWorker.run(DbtTransformationWorker.java:26) ~[io.airbyte-airbyte-commons-worker-0.40.25.jar:?]
		at io.airbyte.workers.temporal.TemporalAttemptExecution.lambda$getWorkerThread$5(TemporalAttemptExecution.java:195) ~[io.airbyte-airbyte-workers-0.40.25.jar:?]
		at java.lang.Thread.run(Thread.java:1589) ~[?:?]
```

**Comment by franviera92** [comment]
this is a serious problem, we need a solution

**Comment by davinchia** [comment]
@GabrielGLevine @franviera92 and everyone else on this thread, unfortunately this has dropped off our roadmap once again in favor of performance and reliability goals. 

I am terrible sorry for the inconvenience. I am happy to help a community member take a stab at this.

Otherwise, one current workaround with Airbyte Kube is to use an external orchestrator to trigger the DBT transformation after the job is run. As an eng, I know it's annoying to set up another system for this and I want us to fix this asap. 

However the current focus is on making reliability/performance improvements (which should benefit everyone on this thread too). Sorry again, and thank you for understanding.

**Comment by fabianofpena** [comment]
Really need this feature! :/

**Comment by redaq** [comment]
We bumped into this wall too. Hope this will be resolved soon!

**Comment by blueyo** [comment]
So the current open source airbyte does not support dbt custom transformation ? I am getting an error related to dbt version. Is that related the discussed above issue? 

> 2023-04-18 13:02:11 dbt > Running: dbt deps --profiles-dir=/data/60/0/transform --project-dir=/data/60/0/transform/git_repo
2023-04-18 13:02:14 dbt > 13:02:14 Encountered an error while reading the project:
2023-04-18 13:02:14 dbt > 13:02:14 ERROR: Runtime Error
2023-04-18 13:02:14 dbt > This version of dbt is not supported with the 'mixpanel' package.
2023-04-18 13:02:14 dbt > Installed version of dbt: =1.0.0
2023-04-18 13:02:14 dbt > Required version of dbt for 'mixpanel': ['>=1.3.0', '<2.0.0']
2023-04-18 13:02:14 dbt > Check the requirements for the 'mixpanel' package, or run dbt again with --no-version-check

**Comment by franviera92** [comment]
> 

Not is run on k8s

**Comment by franviera92** [comment]
> 

@blueyo  Not is run on k8s

**Comment by Grayfados** [comment]
Hi guys... I'm glad to say that I was able to customize the worker as the @davinchia said workaround to run on k8s (I don't remember exactly what slack post was ..) I finished the implementation last week and just entered in Vacation mode kkkkk As soon I return from that I will post here the way that I implemented. Cya!

**Comment by anaselmhamdi** [comment]
Also running into this issue as i was trying to migrate to a kube setup - I'll try to implement a workaround as @davinchia mentioned as well by creating a custom DBT image to supply the transformation

**Comment by Grayfados** [comment]
Hi guys, sorry about my delay. That's the solution that I used to use custom normalization on k8s. 
[https://github.com/airbytehq/airbyte-platform/pull/238](url)

As my main problem is to works with BigqueryDestination I implemented only that. 
The main concern is about the container it self that will run the custom dbt while running the transform-config script.
I created a python script using exclusively standard python libs to work with that.

The generate_profile.py script only deals with Bigquery Destination! So you will need to customize your own use case!

**Comment by glevineLeap** [comment]
I think this is a good feedback opportunity for this group to influence the next version of Normalization: https://github.com/airbytehq/airbyte/issues/25194

**Comment by evantahler** [comment]
> I think this is a good feedback opportunity for this group to influence the next version of Normalization: https://github.com/airbytehq/airbyte/issues/25194

This issue you linked is a bit different. Airbyte is going to stop using dbt internally as the tool to create typed columns in database destinations, and along the way, handle errors better and make saner tables.  

That's unrelated to this feature which lets you run custom dbt transformations after Airbyte has moved the data.

**Comment by 5ylar** [comment]
Any workaround solution for this?

**Comment by skenmy** [comment]
Adding my +1 for this - with this undocumented behaviour I have spent significant time trying to debug this and we are ultimately having to move our deployment into a pet instance to unblock a large migration we have underway.

This is key functionality that breaks when deploying using an [apparently supported](https://docs.airbyte.com/deploying-airbyte/on-kubernetes-via-helm) method and this either needs calling out on the documentation page or fixing as a relative priority!

**Comment by akozichev** [comment]
just hit this issue with Kubernetes.  
As I understand the [Destination V2](https://airbyte.com/blog/introducing-airbyte-destinations-v2-typing-deduping) won't fix it. 
I haven't seen it in action yet but it looks like the "normalisation" with DBT is being removed, whilst custom DBT steps will remain hence the issue will remain.

This wouldn't be a difficult fix if it weren't hidden so far deep inside the worker code.  I couldn't find a way to change the behaviour. Updating CONTAINER_ORCHESTRATOR_IMAGE doesn't help so as providing your own DBT Image. 
The failure happens in airbyte/custom-transformation-prep image. 
I can probably write a mutation webhook in kubernetes to attach an extra volume to all Job pods of airbyte, but this seems to be a bit too much. 
Is there an easy way to use my own container for some of the processing?

**Comment by dramirez-bluon** [comment]
+1 | This feature is really important

**Comment by vatsal-kavida** [comment]
seems like a big issue, +1 for the addition of this feature its a critical one.

**Comment by dhl777** [comment]
+1 | This feature is really important, seems like a big issue, +1 for the addition of this feature its a critical one.

**Comment by cgardens** [comment]
Closed as won't do because we have [deprecated the custom dbt transformation feature](https://github.com/airbytehq/airbyte/discussions/34860).
